core/3d/CCBundle3DData.h

```diff
@@ -39,8 +39,136 @@
 
 #include "3d/CC3DProgramInfo.h"
 
+#include "yasio/detail/byte_buffer.hpp"
+
 NS_CC_BEGIN
 
+class IndexArray
+{
+    static constexpr unsigned int formatToStride(backend::IndexFormat fmt) { return 1 << (int)fmt; }
+    static constexpr backend::IndexFormat strideToFormat(unsigned int stride)
+    {
+        return (backend::IndexFormat)(stride >> 1);
+    }
+
+public:
+    IndexArray() : _stride(formatToStride(backend::IndexFormat::U_SHORT)) {}
+    IndexArray(backend::IndexFormat indexFormat) : _stride(formatToStride(indexFormat)) {}
+
+    IndexArray(std::initializer_list<uint16_t> rhs)
+        : _stride(formatToStride(backend::IndexFormat::U_SHORT)), _buffer(rhs)
+    {}
+    IndexArray(std::initializer_list<uint32_t> rhs, std::true_type /*U_INT*/)
+        : _stride(formatToStride(backend::IndexFormat::U_INT)), _buffer(rhs)
+    {}
+
+    IndexArray(const IndexArray& rhs) : _stride(rhs._stride), _buffer(rhs._buffer) {}
+    IndexArray(IndexArray&& rhs) noexcept : _stride(rhs._stride), _buffer(std::move(rhs._buffer)) {}
+
+    IndexArray& operator=(const IndexArray& rhs)
+    {
+        _stride = rhs._stride;
+        _buffer = rhs._buffer;
+        return *this;
+    }
+    IndexArray& operator=(IndexArray&& rhs) noexcept
+    {
+        this->swap(rhs);
+        return *this;
+    }
+
+    void swap(IndexArray& rhs)
+    {
+        std::swap(_stride, rhs._stride);
+        _buffer.swap(rhs._buffer);
+    }
+
+    void clear() { _buffer.clear(); }
+
+    /** clear with new index format */
+    void clear(backend::IndexFormat format)
+    {
+        clear();
+        _stride = formatToStride(format);
+    }
+
+    /** Pushes back a value. */
+    void push_back(uint32_t val)
+    {
+        assert(_stride == 2 || _stride == 4);
+        _buffer.append_n((uint8_t*)&val, _stride);
+    }
+
+    /** Inserts a list containing unsigned short (uint16_t) data. */
+    void insert(uint8_t* where, std::initializer_list<uint16_t> ilist)
+    {
+        assert(_stride == 2);
+        _buffer.insert(where, (uint8_t*)ilist.begin(), (uint8_t*)ilist.end());
+    }
+
+    /** Inserts a list containing unsigned int (uint32_t) data. */
+    void insert(uint8_t* where, std::initializer_list<uint32_t> ilist, std::true_type)
+    {
+        assert(_stride == 4);
+        _buffer.insert(where, (uint8_t*)ilist.begin(), (uint8_t*)ilist.end());
+    }
+
+    /** Inserts range data. */
+    void insert(uint8_t* where, const void* first, const void* last)
+    {
+        _buffer.insert(where, (const uint8_t*)first, (const uint8_t*)last);
+    }
+
+    template <typename _Ty>
+    _Ty& at(size_t idx)
+    {
+        assert(sizeof(_Ty) == _stride);
+        if (idx < this->size())
+            return (_Ty&)_buffer[idx * sizeof(_Ty)];
+        throw std::out_of_range("IndexArray: out of range!");
+    }
+
+    uint8_t* begin() noexcept { return _buffer.begin(); }
```

Comment:
Because the only safe to iterator IndexArray is it's member function `for_each`, we shoud hide `begin` `end?` functions, then also need do change first parameter of `insert` to `size_t pos`

Reply:
Why would you change `insert` first parameter to `size_t` the insert function of `bute_buffer` takes a uint8_t anyways.

Reply:
I means change uint8_t* pointer to size_t pos, or only provide append function

Reply:
I will provide proposal code later

Reply:
You also wouldn't be able to push from the end since the function `being` `end` have been removed:
![image](https://user-images.githubusercontent.com/45469625/177196290-1048de3c-405b-45d3-bce6-54ee412a24a8.png)

Reply:
Yes, so it could be: `insert(indices.size(), ...`, if we change insert to:
```cpp
insert(size_t off, Iter first, const Iter last);
```

Reply:
You can't though because `_buffer` take a `uint8_t*`

Reply:
Oh okay